### PR TITLE
Sign manifests asynchronously

### LINKED ISSUE #1208: Sign manifests concurrently
**Is your feature request related to a problem? Please describe.**
Manifest signing is serial now. 

**Describe the solution you'd like**
Manifest signing can be sped up by increasing concurrency.

**Additional context**
https://github.com/pulp/pulp_deb/pull/683
https://github.com/pulp/pulp_ansible/blob/main/pulp_ansible/app/tasks/signature.py#L131

### PR DESCRIPTION
**Pull request by lubosmj**
closes #1208

### CODE REVIEW

File: pulp_container/app/tasks/sign.py [hunk 1]
```diff
@@ -81,20 +100,23 @@ def create_signature(manifest, reference, signing_service):
...
+    async with semaphore:
         # download and write file for object storage
+        artifact = await sync_to_async(manifest._artifacts.get)()
         if settings.DEFAULT_FILE_STORAGE != "pulpcore.app.models.storage.FileSystem":
-            manifest_file = tempfile.NamedTemporaryFile(dir=working_directory, delete=False)
-            artifact = manifest._artifacts.get()
-            manifest_file.write(artifact.file.read())
-            manifest_file.flush()
+            async with tempfile.NamedTemporaryFile(dir=".", mode="wb", delete=False) as tf:
+                await tf.write(artifact.file.read())
```
Comment (mdellweg) [inline pulp_container/app/tasks/sign.py:108]:
```suggestion
                await tf.write(await sync_to_async(artifact.file.read)())
```
This should probably not be blocking...

**Comment by lubosmj** [comment]
I cannot really test the changes locally because of multiple permission errors in oci_env. Is someone with a vagrant box volunteering to test the changes? Or, do we trust our CI tests?

**Comment by ipanova** [comment]
I can test, my vagrant box seems to be alive.

File: pulp_container/app/tasks/sign.py (hunk 1 above)
Comment (lubosmj) [reply pulp_container/app/tasks/sign.py]: Based on your latest commit, I am reverting the suggested change:
https://github.com/pulp/pulp_container/pull/1285#discussion_r1196595234

Can you sign this off?

**Comment by ipanova** [comment]
can you update this line https://github.com/pulp/pulp_container/pull/1265/files#diff-94c6436dad7eb9a689d63e5b6e6ceb0d18ddac41ce565c1aa91313c1be8ccfc2R35 we sign manifest lists and children manifests are signed by tag too

File: pulp_container/app/tasks/sign.py [hunk 2]
```diff
@@ -1,8 +1,8 @@
+import asyncio
 import base64
 import hashlib
-import os
-import tempfile
 
+from aiofiles import tempfile
```
Comment (mdellweg) [inline pulp_container/app/tasks/sign.py:5]: Oh... That name is a bit deceiving. But ok.

File: pulp_container/app/tasks/sign.py [hunk 3]
```diff
@@ -47,26 +51,40 @@ def sign(repository_pk, signing_service_pk, reference, tags_list=None):
...
+        async for tag in latest_repo_tags.aiterator():
+            tagged_manifest = tag.tagged_manifest
+            docker_reference = ":".join((reference, tag.name))
+            signature_pk = await create_signature(
+                tagged_manifest, docker_reference, signing_service
+            )
+            added_signatures.append(signature_pk)
+            if tagged_manifest.media_type in MANIFEST_MEDIA_TYPES.LIST:
+                # parse ML and sign per-arches
+                added_signatures = []
```
Comment (mdellweg) [inline pulp_container/app/tasks/sign.py:71]: Why are you clearing that list here?

File: pulp_container/app/tasks/sign.py (hunk 1 above)
Comment (mdellweg) [reply pulp_container/app/tasks/sign.py:107]: Trying to understand what i meant back then. I still think that django storages does not yet know about async. So "artifact.file.read" (which is probably a blocking call,) should not be called from an async context directly, but `sync_to_async` as i suggested.

File: pulp_container/app/tasks/sign.py (hunk 3 above)
Comment (lubosmj) [reply pulp_container/app/tasks/sign.py:71]: I should not. Oopsie woopsie. 👀

File: pulp_container/app/tasks/sign.py [hunk 4]
```diff
@@ -81,20 +99,23 @@ def create_signature(manifest, reference, signing_service):
...
         else:
-            manifest_path = manifest._artifacts.get().file.path
-        sig_path = os.path.join(working_directory, "signature")
+            manifest_path = artifact.file.path
+
+        async with tempfile.NamedTemporaryFile(dir=".", prefix="signature") as tf:
+            sig_path = tf.name
 
-        signed = signing_service.sign(
+        signed = await signing_service.asign(
```
Comment (ipanova) [inline pulp_container/app/tasks/sign.py:118]: where is asign?

File: pulp_container/app/tasks/sign.py [hunk 5]
```diff
@@ -19,6 +20,10 @@
     SIGNATURE_TYPE,
 )
 
+SIGNING_TASKS_COUNTER = 10
```
Comment (ipanova) [inline pulp_container/app/tasks/sign.py:23]: can we expose this to the settings so it is configurable?
I know, this PR then won't be backportable, which is ok given that we don't have now any BZs opened and if we will then backporting without the settings is always an option.

**Review by ipanova — APPROVED** [review]
thanks!

**Pull request merged**